// chrome-extension-scripts/lib/run.js
'use strict';

var _safe = require('colors/safe');

var _safe2 = _interopRequireDefault(_safe);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _webpackDevServer = require('webpack-dev-server');

var _webpackDevServer2 = _interopRequireDefault(_webpackDevServer);

var _easyRequire = require('./utils/easyRequire');

var _easyRequire2 = _interopRequireDefault(_easyRequire);

var _override = require('./webpack/override');

var _override2 = _interopRequireDefault(_override);

var _log = require('./utils/log');

var log = _interopRequireWildcard(_log);

var _shared = require('./shared');

var _webpackConfig = require('./webpack/webpack.config.dev');

var _webpackConfig2 = _interopRequireDefault(_webpackConfig);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

process.env.NODE_ENV = 'development';

// npm


// own


/**
 * Override webpack package files
 * @return {Promise}
 */
function override() {
  return new Promise(function (resolve) {
    resolve((0, _override2.default)());
  });
}

/**
 * Build production release
 *
 * @param  {Object} webpackConfig
 * @return {Promise}
 */
function webpackDevelopment(webpackConfig) {
  return new Promise(function (resolve, reject) {
    return (0, _easyRequire2.default)(function () {

      // TODO move to config
      var host = "0.0.0.0";
      var port = 3001;

      new _webpackDevServer2.default((0, _webpack2.default)(webpackConfig), {
        contentBase: 'https://localhost:3001',
        publicPath: webpackConfig.output.publicPath,
        https: true,
        // lazy: true,
        // watchDelay: 50,
        hot: true,
        // Unfortunately quiet swallows everything even error so it can't be used.
        quiet: false,
        // No info filters only initial compilation it seems.
        noInfo: false,
        // noInfo: true,
        // Remove console.log mess during watch.
        stats: {
          // assets: false,
          colors: true
        }
      }).listen(port, host, function (err, result) {
        if (err) {
          reject('webpack: ' + err);
        } else {
          resolve('Listening at https://' + host + ':' + port);
        }
      });
    });
  });
}

function run(options) {
  override().then((0, _shared.prepareManifest)(options)).then(function (Manifest) {
    return (0, _webpackConfig2.default)(Manifest);
  }).then(webpackDevelopment)
  // Development server ready
  .then(function (message) {
    console.log(_safe2.default.green(message));
  })
  // Some error happened
  .catch(function (error) {
    console.log(_safe2.default.red(error.stack || error));
  });
}

module.exports = run;
