// chrome-extension-scripts/lib/manifest/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();
// import chokidar from 'chokidar'

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _processors = require('./processors');

var _processors2 = _interopRequireDefault(_processors);

var _log = require('../utils/log');

var log = _interopRequireWildcard(_log);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Manifest = function () {
  function Manifest(options) {
    _classCallCheck(this, Manifest);

    this.path = options.manifest;
    this.src = _path2.default.dirname(this.path);
    this.buildPath = options.output;
  }

  _createClass(Manifest, [{
    key: 'run',
    value: function run() {
      this.prepareBuildDir();
      this.processManifest();
      this.writeManifest();
    }

    // watch() {
    //   chokidar.watch(this.path).on('change', this.onChange)
    // }

    // onChange = (event, path) => {
    //   this.processManifest()
    // }

  }, {
    key: 'prepareBuildDir',
    value: function prepareBuildDir() {
      // Prepare clear build
      _fsExtra2.default.removeSync(this.buildPath);
      _fsExtra2.default.mkdirsSync(this.buildPath);
    }
  }, {
    key: 'writeManifest',
    value: function writeManifest() {
      var manifestPath = _path2.default.join(this.buildPath, "manifest.json");
      log.pending('Making \'build/manifest.json\'');
      _fsExtra2.default.writeFileSync(manifestPath, JSON.stringify(this.manifest, null, 2), { encoding: 'utf8' });
      log.done();
    }
  }, {
    key: 'loadManifest',
    value: function loadManifest() {
      return JSON.parse(_fsExtra2.default.readFileSync(this.path, 'utf8'));
    }
  }, {
    key: 'processManifest',
    value: function processManifest() {
      var _this = this;

      this.scripts = [];
      this.manifest = this.loadManifest();

      // Iterate over each processor and process manifest with it
      _processors2.default.forEach(function (processor) {
        _this.applyProcessorResult(processor(_this.manifest, _this));
      });

      return true;
    }
  }, {
    key: 'applyProcessorResult',
    value: function applyProcessorResult() {
      var _this2 = this;

      var _ref = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

      var manifest = _ref.manifest;
      var scripts = _ref.scripts;

      if (manifest) {
        this.manifest = manifest;
      }

      if (scripts) {
        // TODO validace na skripty
        // const pushScriptName = function(scriptName) {
        //   const scriptPath = path.join(paths.src, scriptName)
        //
        //   if(!existsSync(scriptPath)) {
        //     console.warn(colorred(`Missing script ${scriptPath}`))
        //
        //     return
        //   }
        //
        //   if(~scripts.indexOf(scriptName))
        //     return
        //
        //   scripts.push(scriptName)
        // }

        scripts.forEach(function (script) {
          _this2.scripts.push(script);
        });
      }
    }
  }]);

  return Manifest;
}();

exports.default = Manifest;
